### PR TITLE
avoid copies in matrix_cl

### PR DESCRIPTION
**Pull request by t4c1**
## Summary

When a `matrix_cl` is created from data on host, it copies the data to OpenCL device. This is not necessary if the OpenCL device in use uses same memory as host (when the device is CPU or integrated GPU). 

This PR changes `matrix_cl` to instead use the same data if the device uses same memory as host. This is limited to when `matrix_cl` is constructed from a non-const value. Copying from `matrix_cl` to host (`from_matrix_cl`) is not optimized yet, as it would require returning of `Eigen::Map`.

Also added `pinned_matrix` class, which works as an `Eigen::Matrix`, except using OpenCL implementation allocated (likely pinned) memory. When making `matrix_cl` from `pinned_matrix` we should get no-copy behavior on CPU/integrated GPU and a bit faster copying on discrete GPU. `pinned_matrix` is also returned from `ref_type_for_opencl`.

## Tests
TODO: @rok-cesnovar We need to also run OpenCL tests on CPU or an integrated GPU (assuming existing testing uses a discrete GPU?)

## Side Effects
When constructing a `matrix_cl` from a host-allocated object (such as `std::vector` or `Eigen::Matrix`) it is required that this object is not destroyed as long as the `matrix_cl` is in use as it might directly use the memory of that object.

## Release notes

`matrix_cl` now uses host-allocated memory to avoid copies if the OpenCL device uses the same memory as host. Added `pinned_matrix` class.

## Checklist

- [ ] Math issue #(issue number)

- [ ] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [ ] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [ ] the code is written in idiomatic C++ and changes are documented in the doxygen

- [ ] the new changes are tested

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
This is waiting on me to add G.A. OpenCL CPU tests and to do Stan model level benchmarks.

**Comment by SteveBronder** [comment]
:pray: :pray: :pray: this is awesome. @t4c1 do you have docs or a page somewhere you have been reading through on this? I've read the [intel zero copy article](https://software.intel.com/content/www/us/en/develop/articles/getting-the-most-from-opencl-12-how-to-increase-performance-by-minimizing-buffer-copies-on-intel-processor-graphics.html) like a 100 times but always get confused by `clEnqueueMapBuffer()` and `clEnqueueUnmapMemObject()`

**Comment by t4c1** [comment]
No, I was just googling stuff I knew it must exist somewhere in OpenCL.

**Comment by t4c1** [comment]
Hmm I quickly read over the article you linked. It says we will only avoid the copy if the memory is cache line aligned (at least on Intel GPUs), which we can not guarantee in general. Hmm maybe I can improve this a bit.

**Comment by SteveBronder** [comment]
Yeah the exact requirements are weird (I think the article above is for Intel CPUs)

**Comment by t4c1** [comment]
I improved this in two ways:
- Now rvalues are also handled (I found a way to do it without too much complications)
- We now have `pinned_matrix` class.

**Comment by SteveBronder** [comment]
Exciting! I'm trying to finish up the tests for multi_index assign (it's such a wugh) but I'll try to review this either this week or beginning of next (unless @rok-cesnovar you have time to review it earlier)

**Review by SteveBronder — CHANGES_REQUESTED** [review]
Most of this looks good though I'm having trouble following the `pinned_memory` code. Where are you seeing that this is the way to create pinned memory? The OpenCL best practices guide for Nvidia has a more complicated setup for using pinned memory. It may be different now, but I remember AMD and Intel also have different schemes needed for pinned memory as well

http://developer.download.nvidia.com/compute/DevZone/docs/html/OpenCL/doc/OpenCL_Best_Practices_Guide.pdf

File: stan/math/opencl/pinned_matrix.hpp
Comment (t4c1) [reply stan/math/opencl/pinned_matrix.hpp:47]: Whoops, that was used in a previous version of `pinned_matrix`. Not needed anymore

File: stan/math/opencl/pinned_matrix.hpp
Comment (t4c1) [reply stan/math/opencl/pinned_matrix.hpp:45]: We would have to do that if we wanted to use the same buffer object for the device. Instead `matrix_cl`, when constructed from `pinned_matrix` creates a new buffer object (possibly with `CL_USE_HOST_POINTER`) pointing to this buffer.

**Comment by t4c1** [comment]
> The OpenCL best practices guide for Nvidia has a more complicated setup for using pinned memory.

Does it? I went over it again and it seems as exactly the thing `pinned_matrix` is doing. What differences do you see?

**Comment by t4c1** [comment]
There is only one a bit complicated thing in `pinned_matrix` - the need for these private constructors. These are needed because I need to allocate the memory before calling the Map constructor and by default the first thing in constructor is calling superclass constructor (if any is called at all).

**Comment by SteveBronder** [comment]
I think this is good to go but lets wait on the broadcast debug stuff to sort out before we merge this as well

**Comment by rok-cesnovar** [comment]
@SteveBronder the tests now work. I would like to add OpenCL CPU tests before we proceed. We will set that up on the Window machine and once that is enabled we can merge this. Given that its a short test I think there is no reason to not add it.

**Comment by SteveBronder** [comment]
Sounds good to me!

**Comment by rok-cesnovar** [comment]
@serban-nicusor-toptal has setup OpenCL CPU on gelman-group-win2 (Thanks Nic!), we just need the path to the OpenCL.lib on the machine and we can then setup the tests.

**Comment by serban-nicusor-toptal** [comment]
Hey @rok-cesnovar I did a search in the whole `C:\` dir but didn't find any reference to `OpenCL.lib`. Got any idea on where it may be ?

**Comment by t4c1** [comment]
On my PC it is in `C:\Program Files (x86)\Intel\OpenCL SDK\6.1\lib\x64`.

**Comment by serban-nicusor-toptal** [comment]
I've found them in `C:\Program Files (x86)\IntelSWTools\system_studio_2020\OpenCL\sdk\lib\x64` and `C:\Program Files (x86)\IntelSWTools\system_studio_2020\OpenCL\sdk\lib\x86` and also installed the SDK to be sure we're not missing anything.

**Review by SteveBronder — APPROVED** [review]
If the opencl cpu tests are in then this is good to me!

**Comment by bbbales2** [comment]
@t4c1 this is approved but there's a merge conflict

**Comment by rok-cesnovar** [comment]
We are waiting for https://github.com/stan-dev/math/pull/2219 to be green. This should not be merged prior to that.

**Comment by rok-cesnovar** [comment]
Closign this as we will just merge #2219 that includes all these changes anyway. Moved the top-level comment content there as well.

**Pull request closed without merging**
